crates/net/network/src/session/mod.rs

```diff
@@ -589,6 +637,8 @@ async fn authenticate(
     hello: HelloMessage,
     status: Status,
     fork_filter: ForkFilter,
+    blacklisted_peer_ids: Arc<HashSet<PeerId>>,
+    blacklisted_ip: Arc<HashSet<IpAddr>>,
 ) {
     let stream = match direction {
         Direction::Incoming => match ECIESStream::incoming(stream, secret_key).await {
```

Comment:
Should the checks for blacklisted peer id / IP should happen here, rather than after performing the `p2p` handshake? We know the `remote_addr` here, so we can disconnect early if the IP is blacklisted.

In the case of an outgoing connection: We know the remote peer id, so we can immediately disconnect if it is blacklisted.
In the case of an incoming connection: We can perform the ECIES handshake and disconnect if the `stream.remote_id()` is blacklisted.

Reply:
Ideally at the root for every case would be the most efficient. for outgoing, you could block it right at discovery table updates. and you could ip block on binding. However I decided to not due this to keep the code simple as doing filtering on the same kind of case in multiple places imo is a pretty bad practice because it can cause some nice footguns. I'm happy to change it if the consensus is to take a different approach

Reply:
yeah seems more efficient this way and where not spreading out the filtering all over the place